src/build/__init__.py

```diff
@@ -301,13 +379,17 @@ def metadata_path(self, output_directory: PathType) -> str:
         # prepare_metadata hook
         metadata = self.prepare('wheel', output_directory)
         if metadata is not None:
+            match = _WHEEL_NAME_REGEX.match(os.path.basename(metadata))
+            if match:
+                self.project_name = match['distribution']
             return metadata
 
         # fallback to build_wheel hook
         wheel = self.build('wheel', output_directory)
         match = parse_wheel_filename(os.path.basename(wheel))
         if not match:
             raise ValueError('Invalid wheel')
+        self.project_name = match['distribution']
```

Comment:
I'm not very comfortable with mutating the builder in a bunch of different places. If we wanna revalidate the build deps I'd rather we reinitialised the builder by passing the dist name as an explicit parameter.

Reply:
> I'm not very comfortable with mutating the builder in a bunch of different places.

Yeah, this is def a bit messy, unfortunately we don't have a pep517 hook to get our own project name so I'm attempting to pull the name whenever possible.

> If we wanna revalidate the build deps I'd rather we reinitialised the builder by passing the dist name as an explicit parameter.

Well re-validation can often be skipped, if we have a `name` in out `pyproject.yaml` `project` table we can grab it from there to do self cycle checks from the start. It's only when we can't determine our own project name before the build that we need to revalidate it AFAIU.

Reply:
> If we wanna revalidate the build deps I'd rather we reinitialised the builder by passing the dist name as an explicit parameter.

I'm thinking we should probably be validating the normalized project `name` in the various areas where it can potentially be retrieved are all the same and throw an exception if any are different. Right now there doesn't seem to be anything ensuring that a project name set via say the `pyproject.toml` `project` table matches the `wheel`, `sdist` and `dist-info` project names pulled via the regexes.

Reply:
I've implemented project name validation which will throw a `ProjectNameValidationError` exception if at any point a name update is inconsistent with the previous name update. This should ensure that builder project name state changes are done in a way that effectively locks in the normalized project name once one is available.